### ISSUE TITLE
Incorrect observation shape in MLPActorCritic.act in PyTorch implementations

### ISSUE THREAD

**Issue by lubiluk** [openai/spinningup#316]
Hi, in PyTorch docs we can read:
```
torch.nn only supports mini-batches. The entire torch.nn package only supports inputs that are a mini-batch of samples, and not a single sample.

For example, nn.Conv2d will take in a 4D Tensor of nSamples x nChannels x Height x Width.

If you have a single sample, just use input.unsqueeze(0) to add a fake batch dimension.
```
https://pytorch.org/tutorials/beginner/blitz/neural_networks_tutorial.html

In short, it means that we always need a batch dimension. However, in MLPActorCritic.act, we use self.pi with a single observation, without batch dimension. Therefore, the observation shape is incorrect.
I am convinced that we should replace:
```
def act(self, obs, deterministic=False):
        with torch.no_grad():
            a, _ = self.pi(obs, deterministic, False)
            return a.numpy()
```
with
```
def act(self, obs, deterministic=False):
        with torch.no_grad():
            a, _ = self.pi(obs.unsqueeze(dim=0), deterministic, False)
            return a.squeeze(dim=0).numpy()
```

All PyTorch algorithms are affected. I am a bit puzzled why it worked despite this missing dimension. This SO answer gives a potential explanation: https://stackoverflow.com/a/54593741/883843

**Comment by MandiZhao** [comment]
Notice how `act` is used inside the `get_action` functions such as https://github.com/openai/spinningup/blob/038665d62d569055401d91856abb287263096178/spinup/algos/pytorch/td3/td3.py#L267 -- and each observation from the environment is passed to `torch.as_tensor` before getting fed into `act`

**Comment by lubiluk** [comment]
Yes, that changes type from numpy to torch, but doesn't add a batch dimension.

**Comment by MandiZhao** [comment]
Ah I see, the problem is with this inconsistency between `nn.Linear` and `nn.Conv2d`:
Omitting the batch dimension for convolution throws an error (using the example from official doc: https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html#torch.nn.Conv2d
```
>>> m = nn.Conv2d(16, 33, (3, 5), stride=(2, 1), padding=(4, 2), dilation=(3, 1))
>>> input = torch.randn(20, 16, 50, 100)
>>> m(input).shape
torch.Size([20, 33, 26, 100])
>>> input = torch.randn(16, 50, 100)
>>> m(input).shape
Traceback (most recent call last):
RuntimeError: Expected 4-dimensional input for 4-dimensional weight [33, 16, 3, 5], but got 3-dimensional input of size [16, 50, 100] instead
```

But this is not required for linear layer, again using the snippet from https://pytorch.org/docs/stable/generated/torch.nn.Linear.html
```
>>> m = nn.Linear(20, 30)
>>> input = torch.randn(128, 20)
>>> m(input).shape
torch.Size([128, 30])
>>> input = torch.randn(20)
>>> m(input).shape
torch.Size([30])
```

Since spinningup mostly uses feature/state variables for observations -- actor/critic networks are all linear models https://github.com/openai/spinningup/blob/038665d62d569055401d91856abb287263096178/spinup/algos/pytorch/td3/core.py#L28, running the code doesn't throw errors.

**Comment by lubiluk** [comment]
I am aware that it works somehow, even though the documentation says we should always use batch dimension. I would not be sure that it works in a correct way internally though (see the link to SO in my initial post). An interesting thing to test out. 

So, in my opinion this issue brings some confusion and potential source of errors for users of spinningup. When you try something different than Linear then there's an error and effectively you must fork and fix the algo on your end. Therefore I thought it might be a good idea to fix this or at least raise an issue to share this finding with the community.
